marimo/_utils/distributor.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2024 Marimo. All rights reserved.
+from __future__ import annotations
+
+import asyncio
+from threading import Thread
+from typing import Callable, Generic, TypeVar
+
+from marimo import _loggers
+from marimo._utils.disposable import Disposable
+from marimo._utils.typed_connection import TypedConnection
+
+LOGGER = _loggers.marimo_logger()
+
+T = TypeVar("T")
+
+
+class Distributor(Generic[T]):
+    """
+    Used to distribute the response of a multiprocessing Connection to multiple
+    consumers.
+
+    This also handles adding and removing new consumers.
+    """
+
+    def __init__(self, input_connection: TypedConnection[T]) -> None:
+        self.consumers: list[Callable[[T], None]] = []
+        self.input_connection = input_connection
+        self.thread: Thread | None = None
+
+    def add_consumer(self, consumer: Callable[[T], None]) -> Disposable:
+        """Add a consumer to the distributor."""
+        self.consumers.append(consumer)
+
+        def _remove() -> None:
+            if consumer in self.consumers:
+                self.consumers.remove(consumer)
+
+        return Disposable(_remove)
+
+    def _on_change(self) -> None:
+        """Distribute the response to all consumers."""
+        while self.input_connection.poll():
+            try:
+                response = self.input_connection.recv()
+            except (EOFError, StopIteration):
+                break
+            for consumer in self.consumers:
+                consumer(response)
+
+    def start(self) -> Disposable:
+        """Start distributing the response."""
+        asyncio.get_event_loop().add_reader(
```

Comment:
`get_event_loop()` is deprecated starting in Python 3.12, since it silently returns a new event loop if one is not installed. 

The docs recommend replacing this function with `get_running_loop()`, which raises a `RuntimeError` when no event loop is found, though that may make some of your unit tests fail if they don't install an event loop ...

Reply:
i can do this in a followup (give `main` has the same problem), assuming I can get the windows tests to pass without the refactor